nexus/src/app/instance.rs

```diff
@@ -649,6 +608,29 @@ impl super::Nexus {
         self.db_datastore.instance_fetch_with_vmm(opctx, &authz_instance).await
     }
 
+    /// Resizes an instance.
+    pub(crate) async fn instance_resize(
+        &self,
+        opctx: &OpContext,
+        instance_lookup: &lookup::Instance<'_>,
+        new_size: params::InstanceResize,
+    ) -> UpdateResult<InstanceAndActiveVmm> {
+        check_instance_cpu_memory_sizes(new_size.ncpus, new_size.memory)?;
+        let (.., authz_instance) =
+            instance_lookup.lookup_for(authz::Action::Modify).await?;
+
+        self.db_datastore
+            .instance_resize(
+                opctx,
+                InstanceUuid::from_untyped_uuid(authz_instance.id()),
+                new_size.ncpus.into(),
+                new_size.memory.into(),
+            )
+            .await?;
+
+        self.db_datastore.instance_fetch_with_vmm(opctx, &authz_instance).await
```

Comment:
This refetch is kind of sketchy (the instance's shape can change again between the resize query and the fetch). We may not actually need it: since the query is conditional on the instance being halted (i.e. being in `NoVmm`, or possibly in `Failed` once the failed-instance rework lands), there should be no VMM to influence the instance's state, which means we can construct the correct external instance state without having to consult the VMM table to get the active VMM's state. But that's itself kind of sketchy: the rule everywhere else is that you use an `InstanceAndActiveVmm` to get an external-API `Instance` to return, and I think this would be the first exception to that rule.

Another, simpler option would be to have this API return 204 No Content on success. (The caller knows what the updated instance's CPU and memory sizes are because it just set them, and nothing else about the instance is different.) OTOH several other instance routines (stop, reboot, etc.) return the posterior state of the instance, so maybe it's better to be consistent.

Reply:
The 204 No Content approach makes less sense in view of the discussion below (where we've reached a consensus that this API should like other update APIs, which return the posterior states of the things being updated). I think it's probably reasonable to construct an `InstanceAndActiveVmm` in the relevant datastore routine and return it here (though this should have some explanatory comments as @hawkw mentioned above).